### PR TITLE
Migrate to .NET 8

### PR DESCRIPTION
**Pull request by rubo**
## Changes

 - Set the target framework to `net8.0`
 - Suppressed warnings and revised some hacks
 - Added `global.json`
 - Updated workflows
 - Removed obsolete stuff
 - Updated packages

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [x] Other: .NET version upgrade

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

Needs manual testing.

#### Requires documentation update

- [x] Yes
- [ ] No

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

> From now on, Nethermind build artifacts are located in `src/Nethermind/artifacts/bin` directory. 
For instance, for the `release` build configuration, `Nethermind.Runner` artifacts are now in `src/Nethermind/artifacts/bin/Nethemind.Runner/release` while previously they were in `src/Nethermind/Nethemind.Runner/bin/release/net7.0`.

## Remarks

Some changes need to be reconsidered.

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/Collections/LockableConcurrentDictionary.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public static class ConcurrentDictionaryLock<TKey, TValue> where TKey : notnull
     /// <summary>
     /// Delegate that is equivalent of <see cref="ConcurrentDictionary{TKey,TValue}.ReleaseLocks"/>
     /// </summary>
-    private delegate void ReleaseLocks(ConcurrentDictionary<TKey, TValue> dictionary, int fromInclusive, int toExclusive);
+    private delegate void ReleaseLocks(ConcurrentDictionary<TKey, TValue> dictionary, int locksAcquired);
```
Comment (rubo) [inline src/Nethermind/Nethermind.Core/Collections/LockableConcurrentDictionary.cs:25]: This class is an illustrious example of why relying on private APIs is not a good idea. The signature of the method [`ReleaseLocks`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L2108) has been changed in .NET 8 breaking our current implementation.

@LukaszRozmej Do you have any idea on how to make this future-proof?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/Collections/LockableConcurrentDictionary.cs:25]: I don't think we can make it future-proof, the test is there just for this and it worked. So I think we are good.

**Comment by LukaszRozmej** [comment]
Love it, but let's wait for stable.

@kamilchodola @shashankshampi maybe you want to spin up a node or two in the meantime?

**Comment by rubo** [comment]
> Love it, but let's wait for stable.

Absolutely. This is for testing and being ready for the GA.

> @kamilchodola @shashankshampi maybe you want to spin up a node or two in the meantime?

Already run a Mainnet node to compare to the main branch.

File: src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj [hunk 2]
```diff
@@ -13,6 +13,7 @@
     <TieredPGO>true</TieredPGO>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <PublishReadyToRun>true</PublishReadyToRun>
+	<RunAnalyzersDuringBuild>false</RunAnalyzersDuringBuild>
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj:16]: Also use [Dynamically Adapting To Application Sizes](https://maoni0.medium.com/dynamically-adapting-to-application-sizes-2d72fcb6f1ea) for GC?
```suggestion
    <RunAnalyzersDuringBuild>false</RunAnalyzersDuringBuild>
    <GarbageCollectionAdaptationMode>1</GarbageCollectionAdaptationMode>
```

**Review by benaadams — APPROVED** [review]
Also use the new smaller memory GC switch?

**Comment by brbrr** [comment]
Not sure if this is related, but I tried running this branch with sedge and I'm getting a permission error:

```
sedge-execution-client  | 09 Nov 11:47:13 | Step InitDatabase             failed after 218ms System.UnauthorizedAccessException: Access to the path '/nethermind/data/nethermind_db' is denied.
sedge-execution-client  |  ---> System.IO.IOException: Permission denied
sedge-execution-client  |    --- End of inner exception stack trace ---
sedge-execution-client  |    at System.IO.FileSystem.CreateParentsAndDirectory(String fullPath, UnixFileMode unixCreateMode)
sedge-execution-client  |    at System.IO.FileSystem.CreateDirectory(String fullPath, UnixFileMode unixCreateMode)
sedge-execution-client  |    at System.IO.Directory.CreateDirectory(String path)
sedge-execution-client  |    at Nethermind.Db.Rocks.DbOnTheRocks.Init(String basePath, String dbPath, IDbConfig dbConfig, ILogManager logManager, IList`1 columnNames, Boolean deleteOnStart, Nullable`1 sharedCache) in /_/src/Nethermind/Nethermind.Db.Rocks/DbOnTheRocks.cs:line 117
sedge-execution-client  |    at Nethermind.Db.Rocks.DbOnTheRocks..ctor(String basePath, RocksDbSettings rocksDbSettings, IDbConfig dbConfig, ILogManager logManager, IList`1 columnFamilies, Native rocksDbNative, IFileSystem fileSystem, Nullable`1 sharedCache) in /_/src/Nethermind/Nethermind.Db.Rocks/DbOnTheRocks.cs:line 89
sedge-execution-client  |    at Nethermind.Db.Rocks.RocksDbFactory.CreateDb(RocksDbSettings rocksDbSettings) in /_/src/Nethermind/Nethermind.Db.Rocks/RocksDbFactory.cs:line 37
sedge-execution-client  |    at Nethermind.Db.RocksDbInitializer.<>c__DisplayClass15_0.<AddRegisterAction>b__0() in /_/src/Nethermind/Nethermind.Db/RocksDbInitializer.cs:line 55
sedge-execution-client  |    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
sedge-execution-client  | --- End of stack trace from previous location ---
sedge-execution-client  |    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
sedge-execution-client  |    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
sedge-execution-client  | --- End of stack trace from previous location ---
sedge-execution-client  |    at Nethermind.Db.RocksDbInitializer.InitAllAsync() in /_/src/Nethermind/Nethermind.Db/RocksDbInitializer.cs:line 81
sedge-execution-client  |    at Nethermind.Db.StandardDbInitializer.InitStandardDbsAsync(Boolean useReceiptsDb, Boolean useBlobsDb) in /_/src/Nethermind/Nethermind.Db/StandardDbInitializer.cs:line 35
sedge-execution-client  |    at Nethermind.Init.Steps.InitDatabase.Execute(CancellationToken _) in /_/src/Nethermind/Nethermind.Init/Steps/InitDatabase.cs:line 52
sedge-execution-client  |    at Nethermind.Init.Steps.EthereumStepsManager.ExecuteStep(IStep step, StepInfo stepInfo, CancellationToken cancellationToken) in /_/src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:line 156
sedge-execution-client  | 09 Nov 11:47:13 | InitDatabase             complete
```

**Review by brbrr — CHANGES_REQUESTED** [review]
We need to figureout what's up with the file permissions as mentioned in https://github.com/NethermindEth/nethermind/pull/6150#issuecomment-1803686539

**Review by brbrr — APPROVED** [review]
I ran a snap sync on a bunch of node configurations and they all synced fine.

**Comment by smartprogrammer93** [comment]
This ready for review yet? @rubo

**Review by smartprogrammer93 — APPROVED** [review]

**Review by benaadams — APPROVED** [review]

**Review by LukaszRozmej — COMMENTED** [review]
lets goooo!

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**